qiita_db/sql_connection.py

```diff
@@ -630,3 +632,381 @@ def get_temp_queue(self):
+                etext = ('This is likely because the user string `%s` '
+                         'supplied in your configuration file `%s` is '
+                         'incorrect or not an authorized postgres user.' %
+                         (qiita_config.user, qiita_config.conf_fp))
+            elif etype == 'Connection':
+                etext = ('This is likely because postgres isn\'t '
+                         'running. Check that postgres is correctly '
+                         'installed and is running.')
+            else:
+                # we recieved a really unanticipated error with a colon
+                etext = ''
+            ebase = ('An OperationalError with the following message occured'
+                     '\n\n\t%s\n%s For more information, review `INSTALL.md`'
+                     ' in the Qiita installation base directory.')
+            raise RuntimeError(ebase % (e.message, etext))
+
+    def close(self):
+        if self._connection is not None:
+            self._connection.close()
+
+    @contextmanager
+    def _get_cursor(self):
+        """Returns a postgres cursor
+
+        Returns
+        -------
+        psycopg2.cursor
+            The psycopg2 cursor
+
+        Raises
+        ------
+        RuntimeError
+            if the cursor cannot be created
+        """
+        self._open_connection()
+
+        try:
+            with self._connection.cursor(cursor_factory=DictCursor) as cur:
+                yield cur
+        except PostgresError as e:
+            raise RuntimeError("Cannot get postgres cursor: %s" % e)
+
+    def __enter__(self):
+        self._open_connection()
+        self._contexts_entered += 1
+        return self
+
+    def _clean_up(self, exc_type):
+        if exc_type is not None:
+            # An exception occurred during the execution of the transaction
+            # Make sure that we leave the DB w/o any modification
+            self.rollback()
+        elif self._queries:
+            # There are still queries to be executed, execute them
+            # It is safe to use the execute method here, as internally is
+            # wrapped in a try/except and rollbacks in case of failure
+            self.execute()
+            self.commit()
+        elif self._connection.get_transaction_status() != \
+                TRANSACTION_STATUS_IDLE:
+            # There are no queries to be executed, however, the transaction
+            # is still not committed. Commit it so the changes are not lost
+            self.commit()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        # We only need to perform some action if this is the last context
+        # that we are entering
+        if self._contexts_entered == 1:
+            # We need to wrap the entire function in a try/finally because
+            # at the end we need to decrement _contexts_entered
+            try:
+                self._clean_up(exc_type)
+            finally:
```

Comment:
Your solution will mask the exception raised in `self._clean_up` while the current solution will propagate the exception but it will update `self._contexts_entered`. Does it make sense?